Dotty syntax compat

ohze opened the pull request:
<!--
  Are there any relevant issues / PRs / mailing lists discussions?
  Please reference them here - but don't use `fixes` notation.
-->
References #28835

This changeset should be full (binary & source, backward & forward) compatible for scala 2.
Pls see change & message in each commit for more detail.

Due to some bugs in dotty, I add some hacking changes (beside the change in this PR) to a different branch & build that branch in dotty in travis: [![Build Status](https://travis-ci.org/ohze/akka.svg?branch=dotty-wip)](https://travis-ci.org/ohze/akka)

giabao committed:
dotty compatible syntax

+ explicit use `()` when calling/ overriding methods.
  Dotty dropped auto application: http://dotty.epfl.ch/docs/reference/dropped-features/auto-apply.html
+ explicit use `.toString` or use string interpolation instead of using
  `implicit class any2stringadd` due to implicit resolution changes in dotty
+ change `do while` to `while do`

giabao committed:
Only replace all `Props[T]()` to `Props[T]` with all type T. Similar to `TypedProps[T]()`

This changeset is to prepare for dotty compatible:
http://dotty.epfl.ch/docs/reference/dropped-features/auto-apply.html

giabao committed:
Remove SerialVersionUID on traits

https://github.com/lampepfl/dotty/pull/8429

giabao committed:
dotty implicit resolution compat: Add `import akka.actor.actorRef2Scala` to .scala files that use `!` method on ActorRef

giabao committed:
dotty compat: don't use Early Initializers in LightArrayRevolverSchedulerSpec

giabao committed:
make some case class constructors package private instead of `private` to fix dotty compile errors

giabao committed:
Fix Name clash dotty compile error in akka-actor's akka.util.Index

```
error] -- [E120] Duplicate Symbol Error: /Users/thanhbv/ohze/oss/akka/akka-actor/src/main/scala/akka/util/Index.scala:25:18
[error] 25 |    this(mapSize, new Comparator[V] {
[error]    |                  ^
[error]    |Name clash between defined and inherited member:
[error]    |def compare(x$0: T, x$1: T): Int in trait Comparator and
[error]    |def compare(a: V, b: V): Int in anonymous class Object with java.util.Comparator[V] {...} at line 26
[error]    |have the same type after erasure.
```

giabao committed:
Fix dotty Type Mismatch Error in class ActorRef

missing requirement: self type akka.actor.InternalActorRef & akka.actor.ActorRef of class ActorRef
does not conform to self type akka.actor.ActorRefScope of required class InternalActorRef

giabao committed:
explicitly import java.lang.Enum to shadow Enum in dotty

giabao committed:
explicitly declare a method type to fix dotty compile errors

giabao committed:
move TypedMultiMapSpec to scala-2 source dir

We can't use TypedMultiMap in dotty.
http://dotty.epfl.ch/docs/reference/dropped-features/type-projection.html

giabao committed:
TestKit: Use trait parameters instead of Early Initializers for dotty

giabao committed:
fix DockerBindDnsService tests in dotty. Related scala/bug#11924

giabao committed:
fix dotty compile error: non-private constructor Cache in class Cache refers to private class CacheEntry

giabao committed:
fix dotty compile error when write `def copy` for generic case class. lampepfl/dotty#8403

giabao committed:
TestActorRef: don't use Early Initializers (dropped in dotty)

giabao committed:
fix dotty compile error: non-private method unstashed in class StashBufferImpl refers to private class Node

giabao committed:
fix dotty compile error: missing requirement for self type

giabao committed:
dotty compat in testkit & actor-tests: use path-dependent types instead of General Type Projection

giabao committed:
fix type shadow compile error in scala 2.12

giabao committed:
fix dotty compile errors `Cannot find LogSource`

example:
Cannot find LogSource for Class[? <: akka.actor.CoordinatedShutdown]
I found: akka.event.LogSource.fromAnyClass[Nothing]
But method fromAnyClass in object LogSource does not match type akka.event.LogSource[Class[? <: akka.actor.CoordinatedShutdown]]

giabao committed:
re-impl LocalReceptionist using path-dependent type instead of general type projection (dropped in dotty)

This change is backward binary compatible

giabao committed:
prevent dotty to remove unused private fields

ref: lampepfl/dotty#8579

giabao committed:
Fix NullPointerException in DockerBindDnsService.atStartup
when test AsyncDnsResolverIntegrationSpec

creation.warnings() is null and the following check is incorrect:
creation.warnings() should be(null).or(have(size(0)))

Log on travis:
[info] akka.io.dns.AsyncDnsResolverIntegrationSpec *** ABORTED *** (13 seconds, 276 milliseconds)
[info]   java.lang.NullPointerException:
[info]   at org.scalatest.enablers.Size$$anon$2.sizeOf(Size.scala:132)
[info]   at org.scalatest.enablers.Size$$anon$2.sizeOf(Size.scala:132)
[info]   at org.scalatest.matchers.dsl.HaveWord$$anon$6$$anon$1.apply(HaveWord.scala:85)
[info]   at org.scalatest.matchers.MatchersHelper$.orMatchersAndApply(MatchersHelper.scala:161)
[info]   at org.scalatest.matchers.Matcher$$anon$11$$anon$1.apply(Matcher.scala:596)
[info]   at org.scalatest.matchers.should.Matchers$ShouldMethodHelperClass.shouldMatcher(Matchers.scala:6770)
[info]   at org.scalatest.matchers.should.Matchers$AnyShouldWrapper.should(Matchers.scala:6834)
[info]   at akka.io.dns.DockerBindDnsService.atStartup(DockerBindDnsService.scala:80) // line number maybe incorrect

hepin1989 commented:
Very cool

hepin1989 commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -28,35 +28,35 @@ final case class CapturedLogEvent(level: Level, message: String, cause: Option[T
       errorCause: Optional[Throwable],
       marker: Optional[Marker],
-      mdc: java.util.Map[String, Any]) {
+      mdc: java.util.Map[String, Any]) = {
```
With scalafix rule?

ohze replied:
No. dotty compile errors => I just manually change :)

hepin1989 commented on akka-actor/src/main/scala/akka/io/Tcp.scala:
```diff
@@ -106,7 +106,7 @@ object Tcp extends ExtensionId[TcpExt] with ExtensionIdProvider {
    */
   trait Command extends Message with SelectionHandler.HasFailureMessage {
-    def failureMessage = CommandFailed(this)
+    def failureMessage: CommandFailed = CommandFailed(this)
```
with scalfix rule?

ohze replied:
I don't think there will be a scalafix rule for this case.
I don't add explicit type def for all public methods. This case cause dotty compile errors => I manually change.

hepin1989 replied:
but new code will be added then.

ohze replied:
> but new code will be added then.

what? sorry I don't understand.

Blaisorblade replied:
How do you keep Akka building with Dotty, if new contributions are only tested with Scalac and not fixed with Scalafix? (Maybe that’s what @hepin1989 is asking, but it’s a good question in either case). Are you adding Dotty to CI? And would Scalafix help port new contributions to Dotty?

ohze replied:
> Are you adding Dotty to CI?

Yes. I have already described this in the PR description above. Note this image: [![Build Status](https://travis-ci.org/ohze/akka.svg?branch=dotty-wip)](https://travis-ci.org/ohze/akka)

ohze replied:
ok. I know we should use scalafix and create rules if one don't exist.

But for this very specific case `def failureMessage: CommandFailed = CommandFailed(this)`, I don't think we can create a rule to fix this (maybe I am wrong :D). And we don't have many code like this to create a rule.

Rules are good for other cases though, ex "constructor procedure syntax", or "don't use auto application"

hepin1989 commented:
It would be better to extract some changes related to the build definition to a dedicated plugin like `DottySupportPlugin`?

ohze commented:
OK. I will try that idea

giabao committed:
AkkaDisciplinePlugin don't need to depend on Scalafix

ohze commented:
> It would be better to extract some changes related to the build definition to a dedicated plugin like `DottySupportPlugin`.

done

hepin1989 commented:
@ohze Will need a scalafmtAll，I always forgot that too.

raboof commented:
Looks like this needs a copyright header in `project/DottySupportPlugin.scala`

raboof commented on akka-actor/src/main/scala/akka/japi/JavaAPI.scala:
```diff
@@ -83,8 +82,6 @@ object Pair {
  * This class is kept for compatibility, but for future API's please prefer [[akka.japi.function.Creator]].
  */
-@silent("@SerialVersionUID has no effect")
-@SerialVersionUID(1L)
```
While indeed the compiler says this SerialVersionUID 'has no effect', removing it is not strictly speaking binary compatible: `mimaReportBinaryIssues` reports the static field does disappear. Can we be sure that doesn't cause any problems?

ohze replied:
The static field `public static final long serialVersionUID = 1L;` is ONLY generated by scala 2.12.
scala 2.13 don't generate that field.
See also: lampepfl/dotty#8429

ohze replied:
The static field `public static final long serialVersionUID = 1L;` is ONLY generated by scala 2.12.
scala 2.13 don't generate that field.
I think we be sure that doesn't cause any problems.
See also: lampepfl/dotty#8429

raboof commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/LocalReceptionist.scala:
```diff
@@ -60,12 +96,12 @@ private[akka] object LocalReceptionist extends ReceptionistBehaviorProvider {
    */
   private final case class State(
-      services: TypedMultiMap[AbstractServiceKey, Service],
+      services: ServiceMultiMap[ServiceF],
```
Since this is private this change is OK, but our tooling that checks binary compatibility (mima) does flag it for review. Could you add an exclusion (e.g. `ProblemFilters.exclude[Problem]("akka.actor.typed.internal.receptionist.LocalReceptionist*`? You can run `sbt mimaReportBinaryIssues` to find more warnings, and CONTRIBUTING.md has more information on how/where to add the filters.

ohze replied:
done

raboof reviewed (commented):
Pretty cool! I haven't reviewed in depth, but most of the changes I saw seem to be improvements in any case. Could you explain/document somewhere how to run your dotty branch to test?

raboof commented on akka-actor/src/main/scala/akka/actor/Address.scala:
```diff
@@ -23,7 +23,7 @@ import scala.compat.java8.OptionConverters._
  */
 @SerialVersionUID(1L)
-final case class Address private (protocol: String, system: String, host: Option[String], port: Option[Int]) {
+final case class Address private[actor] (protocol: String, system: String, host: Option[String], port: Option[Int]) {
```
The downside if using `private[actor]` here is that we don't have a good way to document that this is not intended to be part of the Java API - however the use of `Option` might be good enough of a hint here...

ohze replied:
Ah. I have just known that `private[akka]` will make the constructor public in java :)
scalac should make it `package private` (`default` in java) :(

patriknw replied:
`private` also makes it public for java, so this doesn't change that aspect
```
  public akka.actor.Address(java.lang.String, java.lang.String, scala.Option<java.lang.String>, scala.Option<java.lang.Object>);
  public akka.actor.Address(java.lang.String, java.lang.String);
  public akka.actor.Address(java.lang.String, java.lang.String, java.lang.String, int);
```

patriknw replied:
so Dotty doesn't allow private constructors at all that can be used from companion, or this is a limitation only for case classes? I assume the latter, otherwise there would be more of these changes.

ohze replied:
`Address`' private constructor cause `copy` and `apply` private
scala/scala#7702
lampepfl/dotty#6671

giabao committed:
fix dotty compile error: Missing type parameter for OutputStreamAppender

giabao committed:
fix dotty compile errors: parentheses are required around the parameter of a lambda

giabao committed:
fix dotty compile errors: Reference to SayHello is ambiguous

giabao committed:
update sbt-scalafix 0.9.13

giabao committed:
scalafix ConstructorProcedureSyntax

```
% sbt
> scalafix github:ohze/scalafix-rules/ConstructorProcedureSyntax
> test:scalafix github:ohze/scalafix-rules/ConstructorProcedureSyntax
> scalafmtAll
> ++ 2.13.1
> scalafix github:ohze/scalafix-rules/ConstructorProcedureSyntax
> test:scalafix github:ohze/scalafix-rules/ConstructorProcedureSyntax
> scalafmtAll
```

giabao committed:
scalafix FinalObject

```
% sbt
> scalafix github:ohze/scalafix-rules/FinalObject
> test:scalafix github:ohze/scalafix-rules/FinalObject
> multi-jvm:scalafix github:ohze/scalafix-rules/FinalObject
> scalafmtAll
> ++ 2.13.1
> scalafix github:ohze/scalafix-rules/FinalObject
> test:scalafix github:ohze/scalafix-rules/FinalObject
> scalafmtAll
```

ohze commented:
> Pretty cool! I haven't reviewed in depth, but most of the changes I saw seem to be improvements in any case. Could you explain/document somewhere how to run your dotty branch to test?

I just tweak the [.travis.yml](https://github.com/ohze/akka/blob/dotty-wip/.travis.yml#L36) file and let travis to do the job when new commits are pushed to `dotty-wip` branch.

You could run the test locally by
```
% git clone git@github.com:ohze/akka.git
% sbt -Dakka.build.scalaVersion=0.24.0-bin-20200324-6cd3a9d-NIGHTLY
> akka-slf4j/test
> akka-actor-tests/test
> akka-actor-typed-tests/compile
...
```
@raboof

ohze commented:
update: I have just split a very large commit (the commit with message `dotty compatible syntax`) to 3 (still large) commits where 2 first commits are output of only using `scalafix` and `scalafmt`.
+ 1 using `scalafix github:ohze/scalafix-rules/ConstructorProcedureSyntax`
+ and 1 using `scalafix github:ohze/scalafix-rules/FinalObject`

patriknw commented:
Thanks for the work here. Looks very promising. We'll try to allocate time for reviewing in more depth next week.

One coordination question. Would it be better to have this merged before the sort imports changes in #28780 or does that not matter much?

giabao committed:
fix dotty compile error in construct: `val x = new { def f = ??? }; x.f`

It's seem dotty will think `x: AnyRef` so `x.f` don't existed

giabao committed:
fix dotty compile error: non-private method apply in object SessionBehavior refers to private class PublishSessionMessage

giabao committed:
fix a dotty compile error when using `implicit class` conversion on `var`

(system : akka.actor.typed.ActorSystem[String]) is not a valid type prefix, since it is not an immutable path

hepin1989 commented on .scalafix.conf:
```diff
@@ -1,6 +1,8 @@
 rules = [
   RemoveUnused
+//  "github:ohze/scalafix-rules/ConstructorProcedureSyntax"
+//  "github:ohze/scalafix-rules/FinalObject"
```
so this was been used locally?

ohze replied:
Yes. It was been used locally but we could just uncomment the lines to run on CI

hepin1989 commented on akka-actor-tests/src/test/scala/akka/actor/SchedulerSpec.scala:
```diff
@@ -692,7 +692,15 @@ class LightArrayRevolverSchedulerSpec extends AkkaSpec(SchedulerSpec.testConfRev
     val prb = TestProbe()
     val tf = system.asInstanceOf[ActorSystemImpl].threadFactory
     val sched =
-      new { @volatile var time = start } with LARS(config.withFallback(system.settings.config), log, tf) {
```
Yes this is been removed from dotty.

hepin1989 commented on akka-actor-tests/src/test/scala/akka/actor/dispatch/DispatcherActorSpec.scala:
```diff
@@ -10,6 +10,7 @@ import java.util.concurrent.{ CountDownLatch, TimeUnit }
 import java.util.concurrent.atomic.AtomicBoolean
 import akka.testkit.AkkaSpec
 import akka.actor.{ Actor, Props }
+import akka.actor.actorRef2Scala
```
https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html

ohze replied:
Hmm. The link is just for reference purpose or you think there is a better way than adding  `import akka.actor.actorRef2Scala`?

hepin1989 commented on akka-actor-tests/src/test/scala/akka/actor/dispatch/PinnedActorSpec.scala:
```diff
@@ -8,6 +8,7 @@ import java.util.concurrent.{ CountDownLatch, TimeUnit }
 
 import akka.testkit._
 import akka.actor.{ Actor, Props }
+import akka.actor.actorRef2Scala
```
https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html

hepin1989 commented on akka-actor-tests/src/test/scala/akka/actor/dungeon/DispatchSpec.scala:
```diff
@@ -5,6 +5,7 @@
 package akka.actor.dungeon
 
 import akka.actor.Actor
+import akka.actor.actorRef2Scala
```
https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html

hepin1989 commented on akka-actor-tests/src/test/scala/akka/dispatch/ControlAwareDispatcherSpec.scala:
```diff
@@ -6,6 +6,7 @@ package akka.dispatch
 
 import akka.testkit.{ AkkaSpec, DefaultTimeout }
 import akka.actor.{ Actor, Props }
+import akka.actor.actorRef2Scala
```
https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala:
```diff
@@ -110,6 +110,7 @@ private[akka] object ActorContextAdapter {
   }
   override private[akka] def internalSpawnMessageAdapter[U](f: U => T, _name: String): ActorRef[U] = {
     val cell = classicContext.asInstanceOf[akka.actor.ActorCell]
+    import akka.actor.actorRef2Scala
```
so this is now
https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html

hepin1989 commented on akka-actor/src/main/scala/akka/actor/Actor.scala:
```diff
@@ -500,7 +500,7 @@ trait Actor {
    * </pre>
    */
-  implicit final val self = context.self //MUST BE A VAL, TRUST ME
+  implicit final val self: ActorRef = context.self //MUST BE A VAL, TRUST ME
```
ah:)

hepin1989 commented on akka-actor/src/main/scala/akka/actor/ActorRef.scala:
```diff
@@ -166,7 +166,7 @@ abstract class ActorRef extends java.lang.Comparable[ActorRef] with Serializable
  * from ActorRef -&gt; ScalaActorRef and back
  */
-trait ScalaActorRef { ref: ActorRef =>
+trait ScalaActorRef { ref: ActorRef with InternalActorRef with ActorRefScope =>
```
is this change safe?

ohze replied:
Yes. Because in old code:
```scala
ScalaActorRef { _: ActorRef => ..}
ActorRef { _: InternalActorRef => ..}
InternalActorRef { _: ActorRefScope => ..}
```
a `ScalaActorRef` must be an `InternalActorRef` and `ActorRefScope` already.
dotty only ensure this by a compile error

hepin1989 commented on akka-actor/src/main/scala/akka/actor/FSM.scala:
```diff
@@ -121,7 +123,7 @@ object FSM {
     private var ref: Option[Cancellable] = _
     private val scheduler = context.system.scheduler
-    private implicit val executionContext = context.dispatcher
+    private implicit val executionContext: ExecutionContextExecutor = context.dispatcher
```
ExecutionContext

ohze replied:
Adding type for implicit members is now performed by scalafix

hepin1989 commented on build.sbt:
```diff
@@ -85,18 +86,20 @@ lazy val actor = akkaModule("akka-actor")
+  .dottySupport()
+  // https://github.com/lampepfl/dotty/issues/8599
+  // TODO remove when the issue is fixed
+  .settings(excludeSourcesIfDotty(Seq("akka/actor/StashJavaAPITestActors.java", "akka/actor/StashJavaAPI.java")))
```
can this from config file?

ohze replied:
I have moved this to another commit that is in [dotty-wip](https://github.com/ohze/akka/blob/dotty-wip/project/DottySupport.scala#L46) branch only (not in `dotty-syntax-compat` branch of this PR)

hepin1989 commented on project/plugins.sbt:
```diff
@@ -25,3 +25,5 @@ addSbtPlugin("de.heikoseeberger" % "sbt-header" % "5.4.0") // for maintenance of
 addSbtPlugin("net.bzzt" % "sbt-reproducible-builds" % "0.24")
 addSbtPlugin("com.dwijnand" % "sbt-dynver" % "4.0.0")
+
+addSbtPlugin("ch.epfl.lamp" % "sbt-dotty" % "0.4.0")
```
Now we have dotty:)

giabao committed:
scalafix: Using ExplicitResultTypes rule to add explicit type for implicit members

Fix by:
```
% sbt
> scalafix
> test:scalafix
> multi-jvm:scalafix
```

giabao committed:
rollback settings used to run scalafix ExplicitResultTypes

giabao committed:
fix compile errors because dotty can't infer types for some vals/ defs

giabao committed:
fix dotty compile error: No ClassTag available for Nothing, Null

giabao committed:
explicitly declaring type for implicit local val / def

ExplicitResultTypes scalafix rule can't fix those cases

giabao committed:
FinalObject fix - for scalafix ignored files

giabao committed:
simplify Dependencies / sslConfigCore sbt setting

giabao committed:
update ssl-config-core 0.4.2

giabao committed:
update scalacheck-1-14 3.1.1.1

giabao committed:
sbt: Add dotty build settings

giabao committed:
sbt: fix an implicitConversions warning

giabao committed:
sbt misc: Add comment about how to import project into IntelliJ with dotty libs

giabao committed:
scalafmtAll

ohze commented:
> Thanks for the work here. Looks very promising. We'll try to allocate time for reviewing in more depth next week.
> 
> One coordination question. Would it be better to have this merged before the sort imports changes in #28780 or does that not matter much?

I think that not matter much because the import clauses added in this PR are in separated lines - so we can just rebase without conflicts if #28780 is merged before this PR

giabao committed:
mima ProblemFilters

giabao committed:
fix dotty compile errors `ambiguous reference`

For example:

Reference to context is ambiguous
it is both defined in object TypedActor
and inherited subsequently in class TypedActor

patriknw commented on akka-actor-tests/src/test/scala/akka/actor/ForwardActorSpec.scala:
```diff
@@ -29,7 +30,7 @@ object ForwardActorSpec {
 class ForwardActorSpec extends AkkaSpec {
   import ForwardActorSpec._
-  implicit val ec = system.dispatcher
+  implicit val ec: ExecutionContextExecutor = system.dispatcher
```
is this needed also if it was `private implicit val`?

if changing this manually I would prrobably use `: ExecutionContext`

ohze replied:
> is this needed also if it was `private implicit val`?

Yes.

> if changing this manually I would prrobably use `: ExecutionContext`

Change by `ExplicitResultTypes` scalafix rule

patriknw commented on build.sbt:
```diff
@@ -500,6 +500,10 @@ def akkaModule(name: String): Project =
+    .settings(
+      addCompilerPlugin(scalafixSemanticdb),
+      scalacOptions += "-Yrangepos"
+    )
```
I don't think we can add this due to https://github.com/lightbend/genjavadoc/issues/220

patriknw commented on build.sbt:
```diff
@@ -500,10 +500,6 @@ def akkaModule(name: String): Project =
-    .settings(
-      addCompilerPlugin(scalafixSemanticdb),
-      scalacOptions += "-Yrangepos"
-    )
```
good, https://github.com/lightbend/genjavadoc/issues/220

patriknw commented on akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:
```diff
@@ -69,12 +69,12 @@ object BehaviorTestKitSpec {
           Behaviors.same
         case SpawnAdapter =>
-          context.spawnMessageAdapter { r: Reproduce =>
+          context.spawnMessageAdapter { (r: Reproduce) =>
```
I'd probably use `context.spawnMessageAdapter[Reproduce] { r =>`, but that's for another PR

ohze replied:
without parentheses will cause error:
```scala
error] -- Error: /Users/thanhbv/ohze/oss/akka/akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:72:51 
[error] 72 |          context.spawnMessageAdapter[Reproduce] { r: Reproduce =>
[error]    |                                                   ^^^^^^^^^^^^
[error]    |parentheses are required around the parameter of a lambda
[error]    |This construct can be rewritten automatically under -language:Scala2Compat -rewrite.
```

patriknw reviewed (commented):
36b4cb0c252dbda417b48f4c4aad7f49dbf4670d is unfortunate. So `ref ! msg` doesn't work without the added import? That's ugly for us (in tests) and Dotty Akka users.

Good that akka.actor.typed.ActorRef.ActorRefOps seems to work without this.

hepin1989 commented:
> [36b4cb0](https://github.com/akka/akka/commit/36b4cb0c252dbda417b48f4c4aad7f49dbf4670d) is unfortunate. So `ref ! msg` doesn't work without the added import? That's ugly for us (in tests) and Dotty Akka users.
> 
> Good that akka.actor.typed.ActorRef.ActorRefOps seems to work without this.

That's why I want to add an `akka.actor.ActorRef.ActorRefOps`....

ohze commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala:
```diff
@@ -128,7 +128,7 @@ private[akka] final class BehaviorTestKitImpl[T](_path: ActorPath, _initialBehav
   }
 
-  override def runOne(): Unit = run(selfInbox.receiveMessage())
+  override def runOne(): Unit = run(selfInbox().receiveMessage())
```
I also think that but the change will cause source backward incompatible.

ohze replied:
maybe break compatibility is acceptable with testkit modules?

ohze replied:
There is a similar problem with `Props.apply` & `TypedProps.apply`.
So I still don't include [this commit](https://github.com/ohze/akka/commit/34d57cd) into this PR although that commit is need for dotty :(
Or the break change here is also acceptable?
(note that it only break source backward compatible, still binary compatible)

renatocaval commented on akka-actor-tests/src/test/scala/akka/actor/DeathWatchSpec.scala:
```diff
@@ -210,19 +210,19 @@ trait DeathWatchSpec { this: AkkaSpec with ImplicitSender with DefaultTimeout =>
 
     "only notify when watching" in {
-      val subject = system.actorOf(Props[EmptyActor]())
+      val subject = system.actorOf(Props[EmptyActor])
```
This change is related to https://github.com/ohze/akka/commit/34d57cd (not included here).

And commented here: https://github.com/akka/akka/pull/28837#discussion_r406270816

where you say:
> although that commit is need for dotty 

Do you really need to remove `()`? I did a quick tests on a tiny dotty project and is does compile fine. 

```scala
case class Props(cls: Class[_])
object Props {
  def apply[T: ClassTag](): Props = Props(implicitly[ClassTag[T]].runtimeClass)
}


val p = Props[String]()
println(p.cls)

```

```shell
[info] running Main 
class java.lang.String
```

renatocaval replied:
Hmm, I think I got why this is needed. 

It's not needed by dotty, but it will be better to remove because then users can't call it as previously `Props[EmptyActor]`. 

The problem is that if we remove we make it source incompatible. We can be sure that there are many projects out there using `Props[EmptyActor]()`.

renatocaval replied:
Maybe we are tackling this from the wrong side. The goal of the PR here, if I understand it correctly, is to make the code source compatible with the Dotty compiler. That said, we should not remove `()` at all. And we should it in the Akka code base in all places that Dotty requires it. 

The day that we provide the binaries for Dotty/Scala3, a user willing to move to Dotty/Scala3 will have to add `()` in all places they omit it before.

renatocaval commented on akka-actor-tests/src/test/scala/akka/event/EventBusSpec.scala:
```diff
@@ -11,39 +11,51 @@ import akka.testkit._
+    Subscriber,
+    BusType <: EventBusSpec.EventBusAux[Event, Classifier, Subscriber]](
+    busName: String,
+    conf: Config = ConfigFactory.empty())
```
yep, the refactoring for type projections will be intrusive. It would be nice to explore some alternatives to reduce the impact on users. However, that won't be zero-impact.

octonato reviewed (commented):
Thanks @ohze for all that work. 

I didn't finish all the files yet. I will continue next week. 
I think we should cut this PR in a few smaller PRs. Making smaller PRs will also help review and push it forward. 

There are a lot of changes that are binary and source compatible. Like, explicit type annotation, explicit usage of `()`. Those we can merge earlier and we should also have some lint rules to avoid introducing then again on new PRs. 

Then we can have a deprecation cycle for the binary-compatible changes that impact users.   

And then we also need to think about the most intrusive source-incompatible changes, like abstract type projection.

hepin1989 commented:
Like first submit the syntax changes and get that be merged and more following up.

ohze commented on akka-actor-tests/src/test/scala/akka/actor/SupervisorHierarchySpec.scala:
```diff
@@ -64,8 +64,8 @@ object SupervisorHierarchySpec {
   case object PongOfDeath
-  final case class Event(msg: Any, identity: Long) { val time: Long = System.nanoTime }
-  final case class ErrorLog(msg: String, log: Vector[Event])
+  final case class MyEvent(msg: Any, identity: Long) { val time: Long = System.nanoTime }
```
If don't rename => dotty compile error in all `case Event(..` in `StressTest` because `FSM` (base trait of `StressTest`) also has `Event` member.

patriknw commented:
This is superseded by smaller PRs, as far as I understand

hepin1989 commented:
@ohze friend ping with 
`
Patrik Nordwall @patriknw Apr 27 20:33
We will release Akka 2.6.5 this week. Probably Thursday.
`
